DataDog/datadog-agent
# Support DNS reverse lookups in windows

### What does this PR do?

Includes support for DNS packet inspection in windows. By implementing PacketSource for windows in packet_source_windows.go, we will get both dns inspection (latency stats) and DNS reverse lookups for free.  

The bulk of this PR is in packet_source_windows.go and driver_interface.go. Driver_interface.go adds a new method, `ReadDNSPacket` which uses overlapped IO to read packet by packet out of the driver. 

### Overlapped IO

Overlapped IO has an unfamiliar interface.   Buffers are created at the start of driver interface and are reused for each read.  An object called a completion port associates the file (in this case a windows.Handle returned from the driver) with several buffers.

* calling ReadFile() takes one of the read buffers, and returns immediately. The kernel is instructed to asynchronously read into the buffer
* a call to GetQueuedCompletionStatus(completionPort) returns the next buffer that has been read into. The userspace process can safely read from the buffer, copy any contents out, and pass it back to the kernel with another call to readfile. Every call to GetQueuedCompletionStatus must  be followed by a call to ReadFile. All this read logic lives in driver_interface.go.            


### Notes on performance

Maxing out UDP connections/second (at 33K cx/second) using dnsstress we saw peak CPU usage 640 millicores, with the spikes being when the connection check ran. And  an 800MB heap.

At a more realistic 1000 requests per second, both heap and CPU increases were not noticeable on the processes page.

After 10 minutes of running at 33K cx/second, we only dropped ~150 packets according to the stats endpoint. This should be well suited to more realistic workloads.  

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

* Spin up a windows host either in virtual box or remotely.  
* Install the datadog agent with network performance monitoring enabled
* browse around in a web browser for a minute
* go to the network page for the host, and group by domain. You should see the domains of the sites you visited.
